ergotree-interpreter/src/eval/savltree.rs

```diff
@@ -0,0 +1,231 @@
+use scorex_crypto_avltree::batch_node::Node;
+use scorex_crypto_avltree::batch_node::NodeHeader;
+use scorex_crypto_avltree::operation::KeyValue;
+use scorex_crypto_avltree::operation::Operation;
+
+use super::EvalError;
+use super::EvalFn;
+
+pub(crate) static INSERT_EVAL_FN: EvalFn = |_env, _ctx, obj, args| {
+    let mut avl_tree_data = match obj {
+        Value::AvlTree(a) => Ok(a),
+        v => Err(EvalError::UnexpectedValue(format!(
+            "AvlTree.insert: expected object of Value::AvlTree got: {0:?}",
+            v
+        ))),
+    }?;
+
+    if !avl_tree_data.tree_flags.insert_allowed() {
+        return Err(EvalError::Misc("AvlTree: Insertions not allowed".into()));
+    }
+
+    let entries = {
+        let v = args
+            .get(0)
+            .cloned()
+            .ok_or_else(|| EvalError::NotFound("AvlTree: eval is missing first arg".to_string()))?;
+        match v {
+            Value::Coll(CollKind::WrappedColl { elem_tpe, items }) => {
+                if elem_tpe
+                    == SType::STuple(STuple::pair(
+                        SType::SColl(Box::new(SType::SByte)),
+                        SType::SColl(Box::new(SType::SByte)),
+                    ))
+                {
+                    let mut tup_items = Vec::with_capacity(items.len());
+                    for i in items {
+                        match i {
+                            Value::Tup(tup) => {
+                                if tup.len() == 2 {
+                                    let first = Bytes::from(
+                                        tup.first().clone().try_extract_into::<Vec<u8>>()?,
+                                    );
+                                    let second = Bytes::from(
+                                        tup.last().clone().try_extract_into::<Vec<u8>>()?,
+                                    );
+                                    tup_items.push((first, second));
+                                } else {
+                                    return Err(EvalError::InvalidResultType);
+                                }
+                            }
+                            _ => unreachable!(),
+                        }
+                    }
+                    tup_items
+                } else {
+                    return Err(EvalError::InvalidResultType);
+                }
+            }
+            _ => return Err(EvalError::InvalidResultType),
+        }
+    };
+
+    let proof = {
+        let v = args.get(1).cloned().ok_or_else(|| {
+            EvalError::NotFound("AvlTree: eval is missing second arg".to_string())
+        })?;
+        Bytes::from(v.try_extract_into::<Vec<u8>>()?)
+    };
+
+    let starting_digest = Bytes::from(avl_tree_data.digest.0.to_vec());
+    let mut bv = BatchAVLVerifier::new(
+        &starting_digest,
+        &proof,
+        AVLTree::new(
+            |digest| Node::LabelOnly(NodeHeader::new(Some(*digest), None)),
```

Comment:
The Rust AVL library has this notion of `Resolver` that's not present in the scala implementation. I just copied what was used in the library example.